### PR TITLE
DKG Contract Client

### PR DESCRIPTION
**Pull request by danuio**
Resolves https://github.com/dapperlabs/flow-go/issues/4427.

Implements the [DKG Contract Client](https://github.com/onflow/flow-go/blob/58bffbec1124f064c9d3aa07b339f3f94ccfbe66/module/dkg.go#L9-L41) interface.

### CODE REVIEW

File: module/dkg.go [hunk 1]
```diff
@@ -3,11 +3,29 @@
...
+
+	sdk "github.com/onflow/flow-go-sdk"
+
 	"github.com/onflow/flow-go/crypto"
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/model/messages"
 )
 
+// DKGSDKClientWrapper is a temporary solution to mocking the `sdk.Client` interface from `flow-go-sdk`
+type DKGSDKClientWrapper interface {
```
Comment (jordanschalm) [inline module/dkg.go:20]: Can we use the same interface for DKG and QC clients?

File: module/dkg/client.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+	tx := sdk.NewTransaction().
+		SetScript(templates.GenerateSendDKGWhiteboardMessageScript(c.env)).
+		SetGasLimit(9999).
+		SetReferenceBlockID(latestBlock.ID).
+		SetProposalKey(c.accountAddress, int(c.accountKeyIndex), account.Keys[int(c.accountKeyIndex)].SequenceNumber).
+		SetPayer(c.accountAddress).
+		AddAuthorizer(c.accountAddress)
+
+	err = tx.AddArgument(cadence.NewString(string(msg.Data)))
```
Comment (jordanschalm) [inline module/dkg/client.go:75]: I think we should convert this (and binary data we're sending to Cadence in general) to hex. It will be easier to debug -- I'm personally not sure what to expect this string to look like with just the type conversion, then sending to Cadence.

File: module/dkg/client.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	if err != nil {
+		return fmt.Errorf("could not add whiteboard dkg message to transaction: %v", err)
```
Comment (jordanschalm) [inline module/dkg/client.go:77]:
```suggestion
		return fmt.Errorf("could not add whiteboard dkg message to transaction: %w", err)
```

File: module/dkg/client.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			return fmt.Errorf("could not get transaction result: %w", err)
+		}
+
+		// if the transaction has expired we skip waiting for seal
+		if result.Status == sdk.TransactionStatusExpired {
+			return fmt.Errorf("broadcast dkg message transaction has expired")
+		}
+
+		// wait 1 second before trying again.
+		time.Sleep(time.Second)
```
Comment (jordanschalm) [inline module/dkg/client.go:106]: can make this a constant

File: module/dkg/client.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+// Messages are returned in the order in which they were broadcast (received
+// and stored in the smart contract)
+func (c *Client) ReadBroadcast(fromIndex uint, referenceBlock flow.Identifier) ([]messages.DKGMessage, error) {
+
+	ctx := context.Background()
+
+	template := templates.GenerateGetDKGLatestWhiteBoardMessagesScript(c.env)
+	dkgMessages, err := c.flowClient.ExecuteScriptAtBlockID(ctx, sdk.Identifier(referenceBlock), template, []cadence.Value{cadence.NewInt(int(fromIndex))})
+	if err != nil {
+		return nil, fmt.Errorf("could not execute read broadcast script")
```
Comment (jordanschalm) [inline module/dkg/client.go:122]:
```suggestion
		return nil, fmt.Errorf("could not execute read broadcast script: %w", err)
```

File: module/dkg/client.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+// represents the group public key and the node's local computation of the
+// public keys for each DKG participant
+func (c *Client) SubmitResult(groupPublicKey crypto.PublicKey, publicKeys []crypto.PublicKey) error {
+
+	ctx := context.Background()
+
+	// get account for given address
+	account, err := c.flowClient.GetAccount(ctx, c.accountAddress, grpc.EmptyCallOption{})
+	if err != nil {
+		return fmt.Errorf("could not get account details: %v", err)
```
Comment (jordanschalm) [inline module/dkg/client.go:139]:
```suggestion
		return fmt.Errorf("could not get account details: %w", err)
```

File: module/dkg/client.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+		SetGasLimit(9999).
+		SetReferenceBlockID(latestBlock.ID).
+		SetProposalKey(c.accountAddress, int(c.accountKeyIndex), account.Keys[int(c.accountKeyIndex)].SequenceNumber).
+		SetPayer(c.accountAddress).
+		AddAuthorizer(c.accountAddress)
+
+	// Note: We need to make sure that we pull the keys out in the same order that
+	// we have done here. Group Public key first followed by the individual public keys
+	finalSubmission := make([]cadence.Value, len(publicKeys))
+	finalSubmission = append(finalSubmission, cadence.NewString(groupPublicKey.String()))
```
Comment (jordanschalm) [inline module/dkg/client.go:159]: 👍 `String` returns the serialized pub key encoded as hex

File: module/dkg/client.go (hunk 6 above)
Comment (jordanschalm) [reply module/dkg/client.go:139]: `%v` doesn't wrap the child error, so it will break error type checking with `errors.As` and `errors.Is` -- you pretty much always want to use `%w` for errors.

File: module/dkg/client.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+		time.Sleep(time.Second)
+	}
+
+	return nil
+}
+
+// submitTx submits a transaction to the flow network and checks that the
+// transaction is signed
+func (c *Client) submitTx(tx *sdk.Transaction) (sdk.Identifier, error) {
+	ctx := context.Background()
```
Comment (jordanschalm) [inline module/dkg/client.go:204]:
```suggestion
func (c *Client) submitTx(ctx context.Context, tx *sdk.Transaction) (sdk.Identifier, error) {
```

We should propagate the same context from the parent scope. The main public `DKGClient` methods should actually probably accept a context as argument from their parent scope as well -- that's my bad not including that in the interface.

File: module/dkg/emulator_client.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	emulator "github.com/onflow/flow-emulator"
+
+	sdk "github.com/onflow/flow-go-sdk"
+)
+
+// EmulatorClient is a wrapper around the emulator to implement the same interface
+// used by the SDK client. Used for testing against the emulator.
+type EmulatorClient struct {
+	blockchain *emulator.Blockchain
+}
```
Comment (jordanschalm) [inline module/dkg/emulator_client.go:21]: We have the same for the `QCClient` as well right? It would be good to merge `feature/epochs` in here and consolidate these.

File: module/dkg/emulator_client.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	result, err := c.blockchain.ExecuteNextTransaction()
+	if err != nil {
+		return err
+	}
+
+	if !result.Succeeded() {
+		return fmt.Errorf("transaction did not succeeded")
+	}
```
Comment (jordanschalm) [inline module/dkg/emulator_client.go:102]: Does `Succeeded` mean that it executed without reverting? To match the semantics of the Access API, we would want to return an error only in the case that the transaction was never executed. The Access API will return no error for transactions submitted that will revert or fail to execute for some reason, since they are executed later.

File: module/dkg/client_test.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	s.dkgAccountKey = accountKey
+	s.dkgSigner = signer
+}
+
+func (s *ClientSuite) TestBroadcast() {
+	err := s.client.Broadcast(messages.NewDKGMessage(0, []byte{}, ""))
+	assert.NoError(s.T(), err)
+}
+
+func (s *ClientSuite) TestReadBroadcast() {
```
Comment (jordanschalm) [inline module/dkg/client_test.go:92]: I think we can expand these tests a bit by testing reading after a broadcast and verifying what we read matches what we wrote. We can also test sending multiple broadcasts and providing a `startIndex` parameter to only read the last `n` messages and verify this functionality.

File: module/dkg/emulator_client.go (hunk 9 above)
Comment (arrivets) [reply module/dkg/emulator_client.go:21]: > We have the same for the `QCClient` as well right? It would be good to merge `feature/epochs` in here and consolidate these.

@jordanschalm I tried merging `feature/epochs` into `feature/dkg`  to see if I could help with this. I am having difficulty getting the appropriate version of the core-contracts module:
https://github.com/onflow/flow-go/blob/512bc5d0d0cea7708092b23aafee9525898dd689/go.mod#L40-L41

The tag `v0.7.2` does not seem to exist (https://github.com/onflow/flow-core-contracts/tags)

Comment (jordanschalm) [reply module/dkg/emulator_client.go:21]: Hey martin, I think the way Go encodes unnamed versions is by prefixing them with a semver version one patch version greater than the last named semver tag in the history. `v0.7.1` is the latest in the history. We are pinning a commit (https://github.com/onflow/flow-core-contracts/commit/ca3212c272a7) but the `v0.7.2` is prefixed for ordering (or something like that). 

What we should be depending on is the head of the https://github.com/onflow/flow-core-contracts/tree/feature/epochs branch in `flow-core-contracts`. When merging we can pin using the latest commit from that branch. Then we'll tag a version when we merge to master.

File: module/emulator/emulator_client.go [hunk 12]
```diff
@@ -90,7 +120,7 @@ func (c *EmulatorClient) Submit(tx *sdk.Transaction) error {
 	}
 
 	if !result.Succeeded() {
-		return fmt.Errorf("transaction did not succeeded")
+		return fmt.Errorf("transaction did not succeeded: %v", result.Error)
```
Comment (jordanschalm) [inline module/emulator/emulator_client.go:123]:
```suggestion
		return fmt.Errorf("transaction did not succeeded: %w", result.Error)
```

File: module/mock/dkgsdk_client_wrapper.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
+// Code generated by mockery v1.0.0. DO NOT EDIT.
```
Comment (jordanschalm) [inline module/mock/dkgsdk_client_wrapper.go:1]: Should this be called just `sdk_client_wrapper.go`? It's shared between both QC and DKG clients right?

File: module/dkg/client.go [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+	}
+}
+
+// Broadcast broadcasts a message to all other nodes participating in the
+// DKG. The message is broadcast by submitting a transaction to the DKG
+// smart contract. An error is returned if the transaction has failed has
+// failed
+func (c *Client) Broadcast(msg model.DKGMessage) error {
+
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply module/dkg/client.go:55]: One other enhancement I think would be worth including is to improve visibility is include some logging in these methods, so we would be able to see if we get stuck in a loop in one of these methods for longer than expected.
We could log:
* when submitting transaction `SubmitTransaction`
* when checking transaction each time `GetTransactionResult` with:
  * time since we started (ie. for this method, time since `Broadcast` was invoked)
  * number of retries

In the happy path case we only send one message in both QC and DKG, so I think it's OK to log these at `info` level

**Review by jordanschalm — COMMENTED** [review]
Lots of minor suggestions and a couple enhancements, but overall looking great. Nice work 🎸

File: module/dkg/client.go [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+		messages = append(messages, flowMsg)
+	}
+
+	return messages, nil
+}
+
+// SubmitResult submits the final public result of the DKG protocol. This
+// represents the group public key and the node's local computation of the
+// public keys for each DKG participant. Serialized pub keys are encoded as hex.
+func (c *Client) SubmitResult(groupPublicKey crypto.PublicKey, publicKeys []crypto.PublicKey) error {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/dkg/client.go:161]: They are ordered on the crypto library level, and the order should not be changed afterwards. Good flag.

[the remark this answers is not in the document]
Comment (tarakby) [reply module/dkg/client.go:161]: Btw, was the contract implemented to see the final submission as an array of strings, or as one long string? 
Fyi, the contract doesn't need to know the details of the array and it can consider the final submission as one string.

**Review by arrivets — APPROVED** [review]

File: module/dkg/client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jordanschalm) [reply module/dkg/client.go:166]: The contract sees [an array of strings](https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowDKG.cdc#L33)

File: module/epochs/qc_voter.go [hunk 16]
```diff
@@ -127,7 +127,7 @@ func (voter *RootQCVoter) Vote(ctx context.Context, epoch protocol.Epoch) error
 		err = voter.client.SubmitVote(ctx, vote)
 		if err != nil {
 			log.Error().Err(err).Msg("could not submit vote - retrying...")
-			continue
+			return fmt.Errorf("could not submit vote: %w", err)
```
Comment (jordanschalm) [inline module/epochs/qc_voter.go:130]: Why this change? This is `QCVoter` not `QCClient`

File: module/dkg/client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply module/dkg/client.go:166]: Great. 
A minor optimization: We can use the first string as the group public key , as done in this PR. The reason is that the contract compares the final submissions going through keys starting from index `0`. The group public key is the one usually being different when final submissions are different. This way the comparison fails early.

**Review by arrivets — CHANGES_REQUESTED** [review]

File: module/dkg/client.go [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+			return fmt.Errorf("could not get transaction result: %w", err)
+		}
+
+		// if the transaction has expired we skip waiting for seal
+		if result.Status == sdk.TransactionStatusExpired {
+			return fmt.Errorf("submit final dkg result transaction has expired")
+		}
+
+		// wait 1 second before trying again.
+		time.Sleep(time.Second)
```
Comment (jordanschalm) [inline module/dkg/client.go:234]: should use the constant here

File: module/dkg/client.go [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply module/dkg/client.go:236]: Looks like we are still missing the `result.Error != nil` check here

File: module/dkg/client.go [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/model/flow"
+	model "github.com/onflow/flow-go/model/messages"
+	"github.com/onflow/flow-go/module"
+)
+
+// Client is a client to the Flow DKG contract. Allows functionality to Broadcast,
+// read a Broadcast and submit the final result of the DKG protocol
+type Client struct {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply module/dkg/client.go:28]: Any thoughts on this? cc @arrivets

File: module/dkg/client.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jordanschalm) [reply module/dkg/client.go]: @danuio Were you thinking to add the logging change in a separate PR? Not seeing any logging added in the second review.

**Review by jordanschalm — APPROVED** [review]
🎸

**Review by arrivets — APPROVED** [review]

File: module/dkg/client.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (danuio) [reply module/dkg/client.go]: Yes, there should be PR out for it today.
